### PR TITLE
Add Error Monitoring to All Environments

### PR DESCRIPTION
**Pull request by ivanELEC**
## Description of change

- Only initialise Sentry if secret for DSN is present Add Active Support breadcrumbs logger
- Include build tag in error details

## Link to relevant ticket
[Add Error Monitoring to All Environments
](https://dsdmoj.atlassian.net/browse/CRM457-194)

### CODE REVIEW

File: gems/laa_multi_step_forms/app/controllers/concerns/error_handling.rb [hunk 1]
```diff
@@ -12,11 +12,10 @@ module ErrorHandling
       # when Errors::ApplicationSubmitted
       #   redirect_to application_submitted_errors_path
       else
-        raise if Rails.application.config.consider_all_requests_local
+        raise unless ENV['RAILS_ENV'] == 'production'
 
+        Sentry.capture_exception(exception) if ENV.fetch('SENTRY_DSN', nil).present?
```
[the remark this answers is not in the document]
Comment (pking-moj) [reply gems/laa_multi_step_forms/app/controllers/concerns/error_handling.rb:17]: Id prefer the guard as it is explicit as to when we are sentry logging.

File: .github/workflows/test-build-deploy.yml [hunk 2]
```diff
@@ -144,16 +144,9 @@ jobs:
...
-      KUBE_NAMESPACE: ${{ secrets.KUBE_NAMESPACE_DEV }}
-
     steps:
       - name: Checkout
-        uses: actions/checkout@v3
-      - name: Update image tag
-        env:
-          ECR_URL: ${{ secrets.ECR_URL_DEV }}
-        run: export IMAGE_TAG=${{ github.sha }} && cat kubectl_deploy/deployment.tpl | envsubst > kubectl_deploy/deployment.yaml
+        uses: actions/checkout@v2
```
[the remark this answers is not in the document]
Comment (ivanELEC) [reply .github/workflows/test-build-deploy.yml]: no, this was an oversight on my side - updated now

File: config/initializers/sentry.rb [hunk 3]
```diff
@@ -1,12 +1,15 @@
...
-  config.dsn = ENV['SENTRY_DSN']
-
-  # Set traces_sample_rate to 1.0 to capture 100%
-  # of transactions for performance monitoring.
-  # We recommend adjusting this value in production.
-  config.traces_sample_rate = 1.0
-  # or
-  config.traces_sampler = lambda do |context|
-    true
+if ENV.fetch('SENTRY_DSN', nil).present?
```
[the remark this answers is not in the document]
Comment (pking-moj) [reply config/initializers/sentry.rb:1]: `ENV.fetch{}` is the preferred way of doing this and Rubocop should be flagging if we are using `ENV[]`

File: gems/laa_multi_step_forms/spec/controllers/error_handling_spec.rb [hunk 4]
```diff
@@ -48,20 +48,21 @@
...
-    before do
-      allow(Rails.application.config).to receive(:consider_all_requests_local).and_return(local)
-    end
-
-    context 'consider_all_requests_local is true' do
-      let(:local) { true }
+    context 'non production RAILS_ENV' do
+      before do
+        ENV['RAILS_ENV'] = 'development'
+      end
```
Comment (dwhenry) [inline gems/laa_multi_step_forms/spec/controllers/error_handling_spec.rb:54]: So this is a change to global state inside the tests suite which wont be reset automatically... this is rarely a good idea.. there are two solutions:

1.
```ruby
around do |spec|
  rails_env = ENV['RAILS_ENV']
  ENV['RAILS_ENV'] = 'development'
  spec.run
  ENV['RAILS_ENV'] = rails_env
end
```
2.
```
before do
  allow(ENV.to receive(:[]).and_call_original
  allow(ENV.to receive(:[]).with('RAILS_ENV').and_return('development')
end
```
I prefer the second as it is a pure mocking solution.. :) 

> NOTE: These code snippets are untested and may not 100% as is

**Review by dwhenry — CHANGES_REQUESTED** [review]
Don't modify global state in tests :(

**Review by dwhenry — APPROVED** [review]

**Review by pking-moj — APPROVED** [review]

**Pull request merged**